feat: sunsetting the cli (#702)
readmeio/oas: README.md

> <p align="center">

Author: Easier to read the changes to this file here: https://github.com/readmeio/oas/tree/feat/sunset-cli

  <a href="https://npm.im/oas">
    <img src="https://raw.githubusercontent.com/readmeio/oas/main/.github/hero.png" alt="oas" />
  </a>
</p>

<p align="center">

>   Comprehensive tooling for working with OpenAPI definitions

Author: I like this way more than "Working with OpenAPI definitions is hard. This makes it easier.".

Reviewer: Idea I just had: thoughts on adding an emoji here?

```suggestion
  Comprehensive tooling for working with OpenAPI definitions 🛠️
```

</p>

<p align="center">
  <a href="https://npm.im/oas"><img src="https://img.shields.io/npm/v/oas.svg?style=for-the-badge" alt="NPM Version"></a>
  <a href="https://npm.im/oas"><img src="https://img.shields.io/node/v/oas.svg?style=for-the-badge" alt="Node Version"></a>
  <a href="https://npm.im/oas"><img src="https://img.shields.io/npm/l/oas.svg?style=for-the-badge" alt="MIT License"></a>
  <a href="https://github.com/readmeio/oas"><img src="https://img.shields.io/github/workflow/status/readmeio/oas/CI.svg?style=for-the-badge" alt="Build status"></a>
</p>

`oas` is the library that we've built at [ReadMe](https://readme.com) for powering everything we do related to [OpenAPI](https://www.openapis.org/); from our [Reference Guides](https://readme.com/documentation), to our [Metrics product](https://readme.com/metrics) or to other in-house tooling like [code generation](https://npm.im/@readme/oas-to-snippet), [request execution](https://npm.im/@readme/oas-to-har), and [SDK code generation](https://api.readme.dev/).

> - [Installation](https://api.readme.dev/docs/installation)

Reviewer: I also kinda like having a header like this but that's a personal preference.
```suggestion

## Table of Contents

- [Installation](https://api.readme.dev/docs/installation)
```

    - [Responses](#responses)
    - [Security](#security)
    - [Specification Extensions](#specification-extensions-1)
  - [Callbacks](#callbacks)
    - [General](#general-2)
  - [Webhooks](#webhooks)
- [FAQ](#faq)

## Installation

```
npm install oas
```

## Usage

> **Note**
> If you need to use this library within a browser you'll likely need to use a bundler like [Webpack](https://webpack.js.org/) or [Rollup](https://rollupjs.org/).

`oas` offers a main `Oas` class, which will be your main entrypoint for using the library.

```js
import Oas from 'oas';
import petstoreSpec from '@readme/oas-examples/3.0/json/petstore.json';

const petstore = new Oas(petstoreSpec);
```

Here the `Oas` constructor takes a JSON API definition (`petstoreSpec`). All API definitions you feed it must be JSON and must be an OpenAPI definition. If you have a YAML or Swagger definition you will need to convert it (our [oas-normalize](https://npm.im/oas-normalize) library can do this for you). And if you're in a CJS or non-`import` environment you can pull the library in with `require('oas').default`.

From here, the following APIs are at your disposal.

### OpenAPI definitions

> Because this library has full TypeScript types and docblocs this README is not intended to be full documentation so consult the individual method docblocks if you need more information on a specific method.

Author: ```suggestion
Because this library has full TypeScript types and docblocks this README is not intended to be full documentation so consult the individual method docblocks if you need more information on a specific method.
```

#### Operations

<!-- prettier-ignore-start -->
| Method | Description |
| :--- | :--- |
| `#findOperation()` | Discover an operation with the current OpenAPI definition that matches a given URL and HTTP method. |
| `#findOperationWithoutMethod()` | Like `oas.findOperation()` but without supplying an HTTP method. |
| `#getOperation()` | Same as `oas.findOperation()` but this returns an instance of the `Operation` class. |
| `#operation()` | Retrieve an instance of the `Operation` or `Webhook` classes for a given path and HTTP method. |
<!-- prettier-ignore-end -->

#### Servers

<!-- prettier-ignore-start -->
| Method | Description |
| :--- | :--- |
| `#defaultVariables()` | Retrieve the default server variables for a specific server URL, while also potentially factoring in user data. You can specify user variable data to the `Oas` constructor. Check out [Using Variables in Documentation](https://docs.readme.com/docs/user-data-options#using-variables-in-documentation) for some background on how we use this. |
| `#replaceUrl()` | Replace a given templated server URL with supplied server variable data. |
| `#splitUrl()` | Chunk out a specific server URL into its individual parts. |

> | `#splitVariables` | Chunk out a given URL and if it matches a server URL in the OpenAPI file, extract the matched server varialbes that are present in the URL. |

Author: ```suggestion
| `#splitVariables` | Chunk out a given URL and if it matches a server URL in the OpenAPI file, extract the matched server variables that are present in the URL. |
```

| `#getTags()` | Retrieve all tags, and [their metadata](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#tagObject), that exist on this operation. |
<!-- prettier-ignore-end -->

#### Callbacks

<!-- prettier-ignore-start -->
| Method | Description |
| :--- | :--- |
| `#getCallback()` | Retrieve a specific callback on this operation. This will return an instance of the `Callback` class. |
| `#getCallbackExamples()` | Retrieve an array of all calback examples that this operation has defined. |
| `#getCallbacks()` | Retrieve all callbacks that this operation has. Similar to `Oas.getPaths()` returning an array of `Operation` instances this will return an array of `Callback`. |
| `#hasCallbacks()` | Determine if this operation has any callbacks defined. |
<!-- prettier-ignore-end -->

#### Parameters

> **Note**
> All parameter accessors here support, and will automatically retrieve and handle, common parameters that may be set at the [path item level](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#pathItemObject).

<!-- prettier-ignore-start -->
| Method | Description |
| :--- | :--- |

> | `#getParameters()` | Retrieve all parameters that maybe used with on this operation. |

Reviewer: ```suggestion
| `#getParameters()` | Retrieve all parameters that may be used with on this operation. |
```

| Method | Description |
| :--- | :--- |
| `#getSecurity()` | Return all security requirements that are applicable for either this operation, or if the operation has none specific to it, then for the entire API. |
| `#getSecurityWithTypes()` | Return a collection of all security schemes applicable to this operation (using `#getSecurity()`), grouped by how the security should be handled (either AND or OR auth requirements). |
| `#prepareSecurity` | Return an object of every security scheme that _can_ be used on this operation, indexed by the type of security scheme it is (eg. `Basic`, `OAuth2`, `APIKey`, etc.). |
<!-- prettier-ignore-end -->

#### Specification Extensions

<!-- prettier-ignore-start -->
| Method | Description |
| :--- | :--- |
| `#getExtension()` | Retrieve a given [specification extension](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#specificationExtensions) if it exists on this operation. |
| `#hasExtension()` | Determine if a given [specification extension](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#specificationExtensions) exists on this operation. |
<!-- prettier-ignore-end -->

> ### Callbacks

Reviewer: Could this section be consolidated with [this one](https://github.com/readmeio/oas/tree/feat/sunset-cli#callbacks)? The hierarchy feels a bit confusing with the placement of this section since callbacks are technically a subset of operations.

Author: I don't think I can do this because callbacks in OAS can actually have their own callbacks so merging the `getIdentifier()` docs into this one I don't think fits because it's only available on the `Callback` instance.

Author: callbacks: not even once

The `Callback` class inherits `Operation` so every API available on instances of `Operation` is available here too. Much like `Operation`, we also support common parameters, summaries, and descriptions that may be set at the [path item level](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#pathItemObject) within a `callbacks` definition.

#### General

<!-- prettier-ignore-start -->
| Method | Description |
| :--- | :--- |
| `#getIdentifier()` | Retrieve the primary identifier of this callback. |
<!-- prettier-ignore-end -->

### Webhooks

> Because our `Webhook` class extensions `Operation`, every API that's available on the [Operation](#operation) class is available on webhooks.

Reviewer: ```suggestion
Because our `Webhook` class extends `Operation`, every API that's available on the [Operation](#operation) class is available on webhooks.
```

Author: lmao. this is what i get for writing documentation at 1am.

## FAQ

#### Can I create an OpenAPI definition with this?

> Though `oas` used to offer functionality related to this it does no longer. If you need an OpenAPI (or Swagger) definition for your API we recommend checking out the language-agnostic [swagger-inline](https://npm.im/swagger-inline), the API editing experience within [ReadMe](https://readme.com), or manually maintaining JSON/YAML files (it sounds worse than it actually is).

Reviewer: dying at this:
> (it sounds worse than it actually is)

<sup>edit: also i would mention <code>swagger-inline</code> last lmao</sup>

```suggestion
Though `oas` used to offer functionality related to this, it does no longer. If you need an OpenAPI (or Swagger) definition for your API we recommend checking out the API editing experience within [ReadMe](https://readme.com), manually maintaining JSON/YAML files (it sounds worse than it actually is), or the language-agnostic [swagger-inline](https://npm.im/swagger-inline).
```
